ethcore/src/miner/miner.rs

```diff
@@ -714,8 +724,14 @@ impl Miner {
 			// | NOTE Code below requires sealing locks.                                |
 			// | Make sure to release the locks before calling that method.             |
 			// --------------------------------------------------------------------------
-			let (block, original_work_hash) = self.prepare_block(client);
-			self.prepare_work(block, original_work_hash);
+			match self.prepare_block(client) {
+				Some((block, original_work_hash)) => {
+					self.prepare_work(block, original_work_hash)
+				},
+				None => {
+					prepare_new = false;
```

Comment:
Isn't that a bit misleading? The function should return `true` if new pending block had to be prepared. And in that case it had to be prepared, but the preparation failed.
Just thinking if it won't break any other assumptions, maybe at least update the function comment.

Reply:
I added `BlockPreparationStatus` which distinguishes three different status.

I think previously I set `prepare_new = false` here is indeed not quite right. We only use the result for `prepare_pending_block` in `prepare_and_update_sealing`. If we don't have to prepare a new block, the function then call `update_sealing`. So if the block had to be prepared but failed, then we shouldn't continue the if statement. (`update_sealing` will also just call `prepare_block` again and early return if it finds out that the conditions does not hold.)